packages/algolia-products/mesh/productFilterInputToAlgoliafacetFiltersInput.ts

```diff
@@ -17,19 +18,35 @@ import { nonNullable } from './utils'
  * https://www.algolia.com/doc/api-reference/api-parameters/facetFilters/
  */
 export function productFilterInputToAlgoliaFacetFiltersInput(
+  settings: AlgoliasettingsResponse,
   filters?: InputMaybe<ProductAttributeFilterInput>,
 ) {
   const filterArray: (string | string[])[] = []
   if (!filters) {
     return []
   }
 
+  const hasVisibility = settings?.attributesForFaceting?.some(
+    (attr) => typeof attr === 'string' && attr.includes('visibility'),
+  )
+
   Object.entries(filters).forEach(([key, value]) => {
     if (isFilterTypeEqual(value)) {
       if (value.in) {
         const values = value.in.filter(nonNullable)
-        if (key === 'category_uid') filterArray.push(values.map((v) => `categoryIds:${atob(v)}`))
-        else filterArray.push(values.map((v) => `${key}:${v}`))
+        if (key === 'category_uid') {
+          filterArray.push(values.map((v) => `categoryIds:${atob(v)}`))
+
+          if (hasVisibility) {
+            filterArray.push(['visibility:Catalog, Search', 'visibility:Catalog'])
+          }
+        } else {
+          filterArray.push(values.map((v) => `${key}:${v}`))
+
+          if (hasVisibility) {
+            filterArray.push(['visibility:Catalog, Search', 'visibility:Search'])
+          }
+        }
```

Comment:
The condition should be:
- If there is a search value, use Catalog, Search + Search
- If there is no search value, use Catalog, Search + Catalog